### PR TITLE
Add JSON/JSONB support

### PR DESCRIPTION
**Pull request by josephglanville**
Somewhat inspired by #83

### CODE REVIEW

File: query.go [hunk 1]
```diff
@@ -284,7 +284,14 @@ func (rows *Rows) Scan(dest ...interface{}) (err error) {
 				rows.Fatal(err)
 			}
 		default:
-			rows.Fatal(fmt.Errorf("Scan cannot decode into %T", d))
+			switch vr.Type().DataType {
```
Comment (josephglanville) [inline query.go]: Is there a better way to do this? It seems hacky but I couldn't find a reasonable way to do it other than matching on interface{} and then falling through into the default case above (which was even more hacky).

**Comment by josephglanville** [comment]
I would like to add JSON/JSONB array support too and look for how these types work in binary encoding.
Any pointers very welcome.

**Comment by josephglanville** [comment]
Will also need a means of disabling the JSON/JSONB tests on older PG versions that don't support either/or those types.

**Comment by jackc** [comment]
I refactored the the Scan method a bit, but it is still complicated. By changing the precedence of JSON decoding vs. normal array decoding we can now decode JSON arrays directly into Go slices.

I added filtering to skip JSON/JSONB tests so it now passes on Travis.

I don't think binary support for JSONB makes sense at this time. The binary format is simply a 1 byte version header, then the document in text.

**Comment by josephglanville** [comment]
Wow awesome.

Yeah I wasn't sure about the binary encoding format, I hadn't looked at the pg sources for how that was done for JSONB yet.

Learnt a few things from how you refactored these changes, thanks very much.

**Pull request merged**